name.abuchen.portfolio/src/name/abuchen/portfolio/snapshot/security/DividendCalculation.java

```diff
@@ -256,4 +265,52 @@ public void visit(CurrencyConverter converter, CalculationLineItem.DividendPayme
         // construct new payment and add it to the list
         payments.add(new Payment(converter, t, getSecurity()));
     }
+
+    /**
+     * calculates the dividend yield on cost for one year. yield is based on the
+     * last dividends and not on an average dividend payment over time for this
+     * security.
+     * 
+     * @param years
+     *            between first and last dividend payments
+     * @return
+     */
+    double calculateYieldOnCost(int years, LocalDate now)
+    {
+        long fifoCostAmount = 0L;
+        for (Payment p : payments)
+        {
+            if (p.fifoCost != null && p.fifoCost.isPositive())
+            {
+                fifoCostAmount = p.fifoCost.getAmount();
+                break;
+            }
+        }
+        if (fifoCostAmount > 0)
+        {
+            double fifoCosts = fifoCostAmount * Values.AmountFraction.factor();
+
+            double diviSum = 0.0;
+            if (years > 1)
+            {
+                LocalDate startTimeframe = now.minusDays(365);
+                for (Payment p : payments)
+                {
+                    if (p.date.isAfter(startTimeframe))
+                    {
+                        diviSum = diviSum + (p.amount.getAmount() * Values.AmountFraction.factor());
+                    }
+                }
+            }
+            else
+            {
+                diviSum = sum.getAmount() * Values.AmountFraction.factor();
+                diviSum = diviSum / years;
+            }
+
+            return diviSum * 100.0 / fifoCosts;
```

Comment:
BTW, this multiplication is the reason for later "magically" divided it by 100. Therefore I would propose just to remove both uses of 100.

Reply:
The reason is at this place is the calculation and the tests are fine with it. The Values class do some magic that is unexpected and not comment/make visible to the developer. Please can you consider to replace the Values class by a new one that do only formatting but no additional calculations? This would give the code a huge improvement in case of understanding what going on here.

Reply:
Vielleicht wäre diese Implementierung besser und nur ein Vorschlag: 
"Optional", um sicherzustellen, dass wir keinen Null-Pointer bekommen, wenn es keine Zahlungen mit FIFO-Kosten gibt.
```
    double calculateYieldOnCost(int years, LocalDate now)
    {
        Optional<Payment> FifoCostAmount = payments.stream() //
                        .filter(p -> p.fifoCost != null && p.fifoCost.isPositive()) //
                        .reduce((r, l) -> l);

        if (FifoCostAmount.isPresent())
        {
            double fifoCosts = FifoCostAmount.get().fifoCost.getAmount() * Values.AmountFraction.factor();

            double diviSum = years > 1 ? payments.stream() //
                            .filter(p -> p.date.isAfter(now.minusDays(365))) //
                            .mapToDouble(p -> p.amount.getAmount() * Values.AmountFraction.factor()) //
                            .sum() //
                            : sum.getAmount() * Values.AmountFraction.factor() / years;

            return diviSum * 100.0 / fifoCosts;
        }

        return 0.0;
    }
```

Reply:
Hi, danke dir.
Mein Code kann aber keine NPE verursachen, da ich das bereits berücksichtigt hatte. Ich habe auch noch mal einen Testfall _testCalculateYieldOnCost_emptyPortfolioShouldNotCauseExceptions_ hinzu gefügt, der demonstriert dass es keine Exceptions gibt.
Deine Lösung mit streams gefällt mir aber. Ich habe den Zwischenteil übernommen. Optional benutze ich nicht innerhalb der eigenen Methode, da man hier seinen eigenen Code in der Hand hat. Wenn es eine zweite Methode gäbe welche die Fifo Kosten berechnet wäre Optional sinnvoll.

@buchen 
@chirlu 
Ich habe an der Stelle gleich mit von 365 Tagen auf 1 Jahr umgestellt. Damit sind nun auch Schaltjahre berücksichtigt.